### PR TITLE
Cluster Readme

### PR DESCRIPTION
**Pull request by stepanyanhayk**
Adding a readme file with instructions on how to run GTSfM on a cluster.

### CODE REVIEW

File: CLUSTER.md [hunk 1]
```diff
@@ -0,0 +1,31 @@
+# GTSFM on CLUSTER
+
+### How to Run GTSfM on a Cluster?
+
+GTSfM uses [SSHCluster](https://docs.dask.org/en/stable/deploying-ssh.html#dask.distributed.SSHCluster) module of [Dask.distributed](https://distributed.dask.org/en/stable/) library to allow utilizing clusters for GTSfM runs. This is a step-by-step guide on how to set up your machines for a successful run on a cluster.
```
Comment (johnwlambert) [inline CLUSTER.md:5]: nit: "uses the SSHCluster module of..."

Comment (johnwlambert) [reply CLUSTER.md:5]: nit: I would say either 
`Gtsfm allows using clusters for SfM execution via the SSHCluster module of Dask`
or
`Gtsfm uses the SSHCluster module of Dask to provide cluster-utilization functionality for SfM execution`

Comment (johnwlambert) [reply CLUSTER.md:5]: I would also say "This readme is a step-by-step guide..."

File: CLUSTER.md [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+1. Enable passwordless SSH between all the workers on the cluster
+    - Log in into a machine
+    - For each of the other workers on the cluster run
+        * ssh-copy-id
```
Comment (johnwlambert) [inline CLUSTER.md:10]: could we put this in code quotes? Do we mean
`ssh-copy-id username@machine_ip_address_of_another_worker`?

Comment (johnwlambert) [reply CLUSTER.md:10]: (the specific bash commands here weren't quite clear to me)

File: CLUSTER.md [hunk 3]
```diff
@@ -0,0 +1,31 @@
+# GTSFM on CLUSTER
+
+### How to Run GTSfM on a Cluster?
+
+GTSfM uses [SSHCluster](https://docs.dask.org/en/stable/deploying-ssh.html#dask.distributed.SSHCluster) module of [Dask.distributed](https://distributed.dask.org/en/stable/) library to allow utilizing clusters for GTSfM runs. This is a step-by-step guide on how to set up your machines for a successful run on a cluster.
+
+1. Enable passwordless SSH between all the workers on the cluster
```
Comment (johnwlambert) [inline CLUSTER.md:7]: nit: should step 1 be to choose which machine will serve as the "scheduler node"?

File: CLUSTER.md [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    - Log in into a machine
+    - For each of the other workers on the cluster run
+        * ssh-copy-id
+        * username@machine_ip_address_of_another_worker
+        * Repeat the above two steps on all machines
+    - Note machines should be able to ssh into themselves passwordless e.g. host1 should be able to ssh into host1.
+2. Clone gtsfm and follow the main readme file to setup the environment on all nodes in the cluster at an identical path
```
Comment (johnwlambert) [inline CLUSTER.md:14]: If we could provide the exact command here, could make this readme even simpler. I think we can assume all machines on the cluster would have to run Linux? does this seem like a fair assumption?

```bash
git clone https://github.com/borglab/gtsfm.git
conda env create -f environment_linux.yml
conda activate gtsfm-v1
```

File: CLUSTER.md [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+3. Log into scheduler again and download the data to scheduler machine
```
Comment (johnwlambert) [inline CLUSTER.md:15]: nit: "Log into the scheduler node again..."

Comment (johnwlambert) [reply CLUSTER.md:15]: (with closing period)

File: CLUSTER.md [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+4. Create a config file listing the cluster workers (example in [gtsfm/configs/cluster.yaml](https://github.com/borglab/gtsfm/blob/master/gtsfm/configs/cluster.yaml))
+5. Run gtsfm with –cluster_config flag enabled, for example
+    - ```
+      python /home/username/gtsfm/gtsfm/runner run_scene_optimizer_colmaploader.py --images_dir /home/username/gtsfm/skydio-32/images/ --config_name sift_front_end.yaml --colmap_files_dirpath /home/hstepanyan3/gtsfm/skydio-32/colmap_crane_mast_32imgs/ --cluster_config cluster.yaml
+      ```
+    - Note that first worker in the cluster.yaml file is the scheduler machine that has the data
```
Comment (johnwlambert) [inline CLUSTER.md:21]: nit: "Note that the first worker in the `cluster.yaml` file must be the scheduler machine where the data is hosted."

File: CLUSTER.md [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+    - Always provide absolute paths for all directories
+6. If you would like to check out the dask dashboard, you will need to do port forwarding from machine to your local computer:
```
Comment (johnwlambert) [inline CLUSTER.md:23]: nit: "...from the scheduler machine to your local computer:"

File: CLUSTER.md [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+    - ```
+      ssh -N -f -L localhost:local_port:localhost:machine_port username@machine_adress
+      ```
+
+7. If you would like to download results from machines to local computer run:
```
Comment (johnwlambert) [inline CLUSTER.md:28]: can be clarify here if the results would appear on multiple machines, or just on the scheduler machine?

**Review by johnwlambert — COMMENTED** [review]
Thanks @stepanyanhayk! A few grammar nits but looks good!

File: CLUSTER.md (hunk 1 above)
Comment (stepanyanhayk) [reply CLUSTER.md:5]: Done.

File: CLUSTER.md (hunk 3 above)
Comment (stepanyanhayk) [reply CLUSTER.md:7]: Done.

File: CLUSTER.md (hunk 2 above)
Comment (stepanyanhayk) [reply CLUSTER.md:10]: Done.

File: CLUSTER.md (hunk 4 above)
Comment (stepanyanhayk) [reply CLUSTER.md:14]: Done.

File: CLUSTER.md (hunk 6 above)
Comment (stepanyanhayk) [reply CLUSTER.md:21]: Done.

File: CLUSTER.md (hunk 7 above)
Comment (stepanyanhayk) [reply CLUSTER.md:23]: Done.

File: CLUSTER.md (hunk 8 above)
Comment (stepanyanhayk) [reply CLUSTER.md:28]: Done.

**Comment by stepanyanhayk** [comment]
Hi, @johnwlambert! I addressed your comments, could you have another look at this PR?

File: CLUSTER.md (hunk 1 above)
Comment (johnwlambert) [reply CLUSTER.md]: Resolved, thanks!

File: CLUSTER.md (hunk 3 above)
Comment (johnwlambert) [reply CLUSTER.md]: Resolved, thanks!

File: CLUSTER.md (hunk 2 above)
Comment (johnwlambert) [reply CLUSTER.md]: Resolved, thanks!

File: CLUSTER.md (hunk 4 above)
Comment (johnwlambert) [reply CLUSTER.md]: Resolved, thanks!

File: CLUSTER.md (hunk 6 above)
Comment (johnwlambert) [reply CLUSTER.md]: Resolved, thanks!

File: CLUSTER.md (hunk 8 above)
Comment (johnwlambert) [reply CLUSTER.md]: Resolved, thanks!

**Review by johnwlambert — APPROVED** [review]

**Pull request merged**